internal/resourcestore/resourcecleaner.go

```diff
@@ -1,33 +1,70 @@
 package resourcestore
 
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/cri-o/cri-o/internal/log"
+)
+
 // A CleanupFunc is a function that cleans up one piece of
 // the associated resource.
-type CleanupFunc func()
+type CleanupFunc func() error
 
 // ResourceCleaner is a structure that tracks
 // how to cleanup a resource.
 // CleanupFuncs can be added to it, and it can be told to
 // Cleanup the resource
 type ResourceCleaner struct {
+	ctx   context.Context
 	funcs []CleanupFunc
 }
 
 // NewResourceCleaner creates a new ResourceCleaner
-func NewResourceCleaner() *ResourceCleaner {
+func NewResourceCleaner(ctx context.Context) *ResourceCleaner {
 	return &ResourceCleaner{
+		ctx:   ctx,
 		funcs: make([]CleanupFunc, 0),
 	}
 }
 
 // Add adds a new CleanupFunc to the ResourceCleaner
-func (r *ResourceCleaner) Add(f func()) {
-	r.funcs = append(r.funcs, CleanupFunc(f))
+func (r *ResourceCleaner) Add(f CleanupFunc) {
+	r.funcs = append(r.funcs, f)
 }
 
 // Cleanup cleans up the resource, running
 // the cleanup funcs in opposite chronological order
 func (r *ResourceCleaner) Cleanup() {
 	for i := len(r.funcs) - 1; i >= 0; i-- {
-		r.funcs[i]()
+		if err := retry(r.funcs[i]); err != nil {
+			log.Errorf(r.ctx, "Retried cleanup function too often, giving up")
+			// TODO: do we wanna return an error here?
```

Comment:
We should bubble up for logging up the stack here.

Reply:
I have updated it to always return the error here